src/lib/communication/DY.Lib.Communication.RequestResponse.fst

```diff
@@ -0,0 +1,206 @@
+
+(*** States ***)
+
+[@@with_bytes bytes]
+type client_send_request = {
+  server:principal;
+  request:bytes;
+  key:bytes
+}
+
+%splice [ps_client_send_request] (gen_parser (`client_send_request))
+%splice [ps_client_send_request_is_well_formed] (gen_is_well_formed_lemma (`client_send_request))
+
+[@@with_bytes bytes]
+type server_receive_request = {
+  request:bytes;
+  key:bytes
+}
+
+%splice [ps_server_receive_request] (gen_parser (`server_receive_request))
+%splice [ps_server_receive_request_is_well_formed] (gen_is_well_formed_lemma (`server_receive_request))
+
+[@@with_bytes bytes]
+type client_receive_response = {
+  server:principal;
+  response:bytes;
+  key:bytes
+}
+
+%splice [ps_client_receive_response] (gen_parser (`client_receive_response))
+%splice [ps_client_receive_response_is_well_formed] (gen_is_well_formed_lemma (`client_receive_response))
+
+[@@with_bytes bytes]
+type communication_states =
+  | ClientSendRequest: client_send_request -> communication_states
+  | ServerReceiveRequest: server_receive_request -> communication_states
+  | ClientReceiveResponse: client_receive_response -> communication_states
+
+#push-options "--ifuel 1"
+%splice [ps_communication_states] (gen_parser (`communication_states))
+%splice [ps_communication_states_is_well_formed] (gen_is_well_formed_lemma (`communication_states))
+#pop-options
+
+instance parseable_serializeable_bytes_communication_states: parseable_serializeable bytes communication_states
+  = mk_parseable_serializeable ps_communication_states
+
+instance local_state_communication_layer_session: local_state communication_states = {
+  tag = "DY.Lib.Communication.State.RequestResponse";
+  format = parseable_serializeable_bytes_communication_states;
+}
+
+[@@with_bytes bytes]
+type communication_reqres_event =
+  | CommClientSendRequest: client:principal -> server:principal -> request:bytes -> key:bytes -> communication_reqres_event
+  | CommServerReceiveRequest: server:principal -> request:bytes -> key:bytes -> communication_reqres_event
+  | CommServerSendResponse: server:principal -> request:bytes -> response:bytes -> communication_reqres_event
+  | CommClientReceiveResponse: client:principal -> server:principal -> response:bytes -> key:bytes -> communication_reqres_event
+
+#push-options "--ifuel 1"
+%splice [ps_communication_reqres_event] (gen_parser (`communication_reqres_event))
+%splice [ps_communication_reqres_event_is_well_formed] (gen_is_well_formed_lemma (`communication_reqres_event))
+#pop-options
+
+instance event_communication_reqres_event: event communication_reqres_event = {
+  tag = "DY.Lib.Communication.Event.RequestResponse";
+  format = mk_parseable_serializeable ps_communication_reqres_event;
+}
+
+(*** Layer Setup ***)
+
+val comm_layer_aead_tag: string
+let comm_layer_aead_tag = "DY.Lib.Communication.Aead.Key"
+
+[@@with_bytes bytes]
+type comm_meta_data = {
+  key:bytes;
+  server:principal;
+  sid:state_id;
+  request:bytes;
+}
```

Comment:
If I understand correctly, this type is produced by `send_request` or `receive_request` and then used by `send_response` or `receive_response`? And the attacker should pass it around these functions?

If so, I wouldn't put `bytes` in there, such as `key` or `request`, because it then means the attacker has in its possession `bytes` that it isn't supposed to know. These `bytes` should only be stored in a state (which is actually the case, in `communication_states`)

Reply:
However if we do not make these functions directly available to the attacker and they are wrapped together in another function (e.g. written by users of communication layer) then the comment above doesn't apply.

In that case the comment is then that there seems to be duplication between the `bytes` in `comm_meta_data` and the `bytes` in `communication_states`).

Reply:
(see #106 )